File: src/nack.php
<?php

function parse_nmap_output($stdout)
{
    $result = [];
    $lines = explode("\n", $stdout);
    $current_ip_address = "";

    foreach ($lines as $line) {
        if (strpos($line, 'Nmap scan report for') !== false) {
            $current_ip_address = explode(' ', $line)[4];
            $result[$current_ip_address] = [];
        } elseif ((strpos($line, 'open') !== false || strpos($line, 'filtered') !== false || strpos($line, 'closed') !== false) && strpos($line, '/') !== false) {
            $split_line = explode(' ', $line);
            $port_protocol = explode('/', $split_line[0]);
            $port = (int) $port_protocol[0];
            $protocol = $port_protocol[1];
            $state = $split_line[1];
            $service = $split_line[2];

            $result[$current_ip_address][$port] = [
                'state' => $state,
                'protocol' => $protocol,
                'service' => $service,
            ];
        }
    }

    return $result;
}

$argc = count($argv);

if ($argc !== 2) {
    fwrite(STDERR, "Please provide an IP address, range of IPs, or a domain as a command-line argument.\n");
    exit(1);
}

$target = $argv[1];
$command = "nmap -p- {$target}";

$descriptorspec = [
    0 => ["pipe", "r"],
    1 => ["pipe", "w"],
    2 => ["pipe", "w"],
];

$process = proc_open($command, $descriptorspec, $pipes);

if (is_resource($process)) {
    fclose($pipes[0]);
    $stdout = stream_get_contents($pipes[1]);
    fclose($pipes[1]);
    $stderr = stream_get_contents($pipes[2]);
    fclose($pipes[2]);

    $return_value = proc_close($process);

    if ($return_value !== 0) {
        fwrite(STDERR, "exec error: {$stderr}\n");
        exit(1);
    }

    $open_ports = parse_nmap_output($stdout);
    echo json_encode($open_ports) . PHP_EOL;
}

File: src/nack.lua
local json = require 'dkjson'

function parse_nmap_output(stdout)
    local result = {}
    local lines = {}
    for line in stdout:gmatch("[^\r\n]+") do
        table.insert(lines, line)
    end
    local current_ip_address = ""

    for _, line in ipairs(lines) do
        if string.find(line, "Nmap scan report for") then
            current_ip_address = string.match(line, "%S+%s+%S+%s+%S+%s+(%S+)")
            result[current_ip_address] = {}
        elseif (string.find(line, "open") or string.find(line, "filtered") or string.find(line, "closed")) and string.find(line, "/") then
            local split_line = {}
            for word in line:gmatch("%S+") do
                table.insert(split_line, word)
            end
            local port_protocol = {}
            for item in split_line[1]:gmatch("[^/]+") do
                table.insert(port_protocol, item)
            end
            local port = tonumber(port_protocol[1])
            local protocol = port_protocol[2]
            local state = split_line[2]
            local service = split_line[3]

            result[current_ip_address][port] = {
                state = state,
                protocol = protocol,
                service = service
            }
        end
    end

    return result
end

local args = {...}
if #args ~= 1 then
    print("Please provide an IP address, range of IPs, or a domain as a command-line argument.")
    os.exit(1)
end

local target = args[1]
local command = "nmap -p- " .. target

local process = io.popen(command, "r")
local stdout = process:read("*a")
process:close()

local open_ports = parse_nmap_output(stdout)
print(json.encode(open_ports, {indent = true}))
